### ISSUE TITLE
Fix S4070 FP/FN: False positive when using left shift operator

### ISSUE THREAD

**Issue by vbreuss** [SonarSource/sonar-dotnet#7505]
### Description
`Remove the 'FlagsAttribute' from this enum` is displayed, although the enum only uses power of two values.

### Repro steps
Define the following enum:
```csharp
[Flags]
public enum EnumExamples
{
  Bit1 = 1 << 0,
  Bit2 = 1 << 1,
  Bit3 = 1 << 2
}
```

Even though the values are clearly power of two, Sonarcloud raises S4070

### Expected behavior

No issue is raised, because all enum values are a power of two.

### Actual behavior

`Remove the 'FlagsAttribute' from this enum` is displayed:
![image](https://github.com/SonarSource/sonar-dotnet/assets/3438234/f4e9dc14-a712-47ac-b980-a2017a74c774)

### Known workarounds

If providing the values as `1`, `2`, `4` no issue is generated, but especially with large values, this gets really hard to read.

**Comment by gregory-paidis-sonarsource** [comment]
Hey @vbreuss , thanks for raising this.

I tried reproducing it locally, but I did not succeed. For me, it raises as intended. 
Also, looking at the codebase, the case seems to be covered. 

You can see this in action in [this reproducer SonarCloud repo](https://sonarcloud.io/code?id=greggy42_s4070-fp-repro&selected=greggy42_s4070-fp-repro%3AProgram.cs) I just made.

Notice that for `Good1` and `Good2` there is no issue raised for S4070, whereas for `Bad` it raises as expected.

Could you maybe share a bit more about your scenario? 
If your project is open-source, feel free to link it :)

**Comment by vbreuss** [comment]
Thanks for the feedback @gregory-paidis-sonarsource.

I fear I mixed two things:
- We had a lot of these issues with the left shift operator in our previous SonarQube instance where we all marked them as "False Positive"
- Now we had the same issue again in SonarCloud and I raised the issue without having a more detailed look.
Sorry for the inconvenience!

After a more detailed analysis, the problem we have is not the bit-shift itself, but that we mix special bits with new ones. 
Consider the following example where we have two types of servers and use bit 0 to distinguish the two types:
```csharp
[Flags]
public enum ServerTypes
{
  IsRemote = 1 << 0,
  LocalUserServer = 1 << 1,
  RemoteUserServer = LocalUserServer | IsRemote,
  // Some other servers that can run local or remote
  LicenseServer = 1 << 5 | IsRemote
}
```

This raises [S4070](https://sonarsource.github.io/rspec/#/rspec/S4070/csharp) for the `LicenseServer` even though it is a combination of power of two and another defined value.

*Should this really be considered an invalid `Flags` enum?*

**Comment by gregory-paidis-sonarsource** [comment]
Hey again @vbreuss . Thanks for clarifying!

I understand your usecase a bit better now. I will add this to our dedicated backlog for FPs, where it will be picked-up during some upcoming hardening sprint.

Cheers!

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1722](https://sonarsource.atlassian.net/browse/NET-1722)

[NET-1722]: https://sonarsource.atlassian.net/browse/NET-1722?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
